Thesis: Integration Chemical Knowledge into GNN

aditya0by0 opened the pull request:
Primary Tasks:
- Rule-based Functional Group detection including rings
- Augmented GNN with artificial nodes for each Functional group
- GAT Message passing mechanism

aditya0by0 committed:
fg rule-based detection algo from FARM

aditya0by0 committed:
Update .pre-commit-config.yaml

aditya0by0 committed:
Update .gitignore

aditya0by0 committed:
remove redundant code functions

aditya0by0 committed:
reader: add fg augment reader

aditya0by0 committed:
reader: add check for graph functional group

aditya0by0 committed:
reader: add ring size to node feature

aditya0by0 committed:
update logic for node features to be processed by properties

aditya0by0 committed:
move fg constants to seperate file

aditya0by0 committed:
Elements: re-align

aditya0by0 commented on chebai_graph/preprocessing/reader.py:
z

aditya0by0 commented:
@sfluegel05 ,  
I was reviewing the `GraphPropertyReader` class and  I have a doubt  in the `_read_data` method, the graph seems to be constructed as **directed**, since the `edge_index` is populated only in one direction (i.e., `[begin_atom_idx] -> [end_atom_idx]`).

```python
edge_index = torch.tensor(
    [
        [bond.GetBeginAtomIdx() for bond in mol.GetBonds()],
        [bond.GetEndAtomIdx() for bond in mol.GetBonds()],
    ]
)
```

Just want to make sure that — is this an **intentional decision** to work with directed graphs for molecules? Or is this converted to an **undirected** graph later in the pipeline which I might have missed?

Since chemical bonds are usually considered undirected in molecular graph models, so I had this doubt.

aditya0by0 committed:
properties: add base class for augmented bond property

aditya0by0 committed:
move augmented properties to utils

aditya0by0 committed:
add bond level prop for augmented bond

aditya0by0 commented:
@sfluegel05, I'm currently encountering another issue while running the graph repository. Are you experiencing the same error  while running it from your end as well?

```bash
G:\anaconda3\envs\gnn\python.exe G:\github-aditya0by0\python-chebai\chebai fit --trainer=configs/training/default_trainer.yml --trainer.callbacks=configs/training/default_callbacks.yml --model=G:/github-aditya0by0/python-chebai-graph/configs/model/gnn_res_gated.yml --model.train_metrics=configs/metrics/micro-macro-f1.yml --model.test_metrics=configs/metrics/micro-macro-f1.yml --model.val_metrics=configs/metrics/micro-macro-f1.yml --data=G:/github-aditya0by0/python-chebai-graph/configs/data/chebi50_graph.yml --model.criterion=configs/loss/bce.yml --data.init_args.batch_size=40 --data.init_args.num_workers=12 --model.pass_loss_kwargs=false --data.init_args.chebi_version=227 --trainer.min_epochs=1 --trainer.max_epochs=1 

error: Parser key "data":
  Does not validate against any of the Union subtypes
  Subtypes: [<class 'NoneType'>, <class 'lightning.pytorch.core.datamodule.LightningDataModule'>]
  Errors:
    - Expected a <class 'NoneType'>
    - metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
  Given value type: <class 'dict'>
  Given value: {'class_path': 'chebai_graph.preprocessing.datasets.chebi.ChEBI50GraphData'}
```

aditya0by0 committed:
add fg reader

aditya0by0 committed:
AtomFG prop reads from fgreader

aditya0by0 committed:
move properties to properties dir

aditya0by0 committed:
data config for augmented graph

aditya0by0 committed:
right molecular prop import

aditya0by0 committed:
move prop constants to props dir

aditya0by0 committed:
make data dir a python dir

aditya0by0 committed:
Update .gitignore

aditya0by0 committed:
remove properties imports in preprocessing init

aditya0by0 committed:
resolve metaclass error

aditya0by0 committed:
move rule based reader to class init to avoid circular import

aditya0by0 committed:
right order of import in properties init

aditya0by0 committed:
remove graph fg check in augmentor reader

aditya0by0 committed:
right order of imports

aditya0by0 committed:
make rings as a Functional Group

aditya0by0 committed:
utility to visualize augmented molecule

aditya0by0 committed:
rename fg_graph edge constant

aditya0by0 committed:
add diff color to diff edge type

aditya0by0 committed:
Update visualize_augmented_molecule.py

aditya0by0 committed:
fix edge color mismatch bug

aditya0by0 committed:
better position alignment of the nodes

aditya0by0 committed:
add CLI

aditya0by0 committed:
plot based on given plot type

aditya0by0 committed:
Update .pre-commit-config.yaml

aditya0by0 committed:
add 3d plot visualization

aditya0by0 committed:
pre-commit formatting

aditya0by0 committed:
ploty lib is a soft requirement, use dynamic import error

aditya0by0 committed:
modularize plotting code

aditya0by0 committed:
restructure and add docstring, typehints

aditya0by0 committed:
proper namespace to avoid future conflicts

aditya0by0 committed:
seperate file for augmented reader

aditya0by0 committed:
abstract base class for augmentor readers

aditya0by0 committed:
for 3d plot hover info to be id of the node

aditya0by0 committed:
logic change to detect FG related to rings

aditya0by0 commented:
@sfluegel05, Please find the script to visualize the augmented molecule using the below script and parameters 

```bash
chebai_graph\preprocessing\utils\visualize_augmented_molecule.py plot --smiles="OC(=O)c1ccccc1O" --plot_type="3d" 
```

Some complex SMILES are mentioned in the comments at the end of script

aditya0by0 committed:
restructure reader and add docstrings + typehints

aditya0by0 committed:
rename to `fg_aware_rule_based` algorithm

aditya0by0 committed:
Create .gitignore

aditya0by0 committed:
update precommit + github action

aditya0by0 committed:
pre-commit format files

aditya0by0 committed:
change graph from directed to UNDIRECTED

aditya0by0 committed:
properties error fix

aditya0by0 committed:
fix props error

aditya0by0 committed:
make augment prop a subclass of atom, bond prop

aditya0by0 committed:
remove rule based reader

aditya0by0 committed:
add vscode to ignore

aditya0by0 committed:
add data class for augmentation

aditya0by0 committed:
skip if reader return None, using walrus operator in python >=3.8

aditya0by0 committed:
molecular prop also accepts dict in case of augmentation

aditya0by0 committed:
update augmented reader for wildcard smiles

aditya0by0 committed:
fix to cal right number of nodes

aditya0by0 committed:
assert if num of prop values is equal to num of nodes/edges

aditya0by0 committed:
take value from reader if not None

aditya0by0 committed:
add test data

aditya0by0 committed:
edge_features should be calculated after undirected graph

aditya0by0 committed:
directed edge which form an un-dir edge should be adjancent

aditya0by0 committed:
add test for GraphPropertyReader

aditya0by0 committed:
add gt test data for aspirin

aditya0by0 committed:
Update test_data.py

aditya0by0 committed:
dir edges of undir should be adjacent

aditya0by0 committed:
assert checks for aug reader

aditya0by0 committed:
add more graph test

aditya0by0 committed:
first src to tgt edges then tgt to src

- instead of using adjacent directed edge, this one is better approach since we can stack edge attributes generated later without any further logic to rearrange edge_attr

aditya0by0 committed:
add test for duplicate directed edges

aditya0by0 committed:
restore import

aditya0by0 committed:
concat edge attr for undirected graph

aditya0by0 committed:
concat prop values instead of edge_attr

aditya0by0 committed:
Merge branch 'dev' into fix/directed-to-undirected-graph

aditya0by0 committed:
edge_index: first src to tgt edges, then tgt to src edges

aditya0by0 committed:
Merge branch 'dev' into fix/directed-to-undirected-graph

aditya0by0 committed:
inherit from ChebiOverX

instead of Base data module, as `load_processed_data_from_file` method used in this class is available in Dynamic dataset class

aditya0by0 committed:
add augmented props to config

aditya0by0 committed:
gat wrapper

aditya0by0 committed:
add linear layers to gat

aditya0by0 committed:
gat ffn fix

aditya0by0 committed:
`nan_to_num` numpy2.x compatibility

fix for https://github.com/ChEB-AI/python-chebai-graph/issues/10

aditya0by0 committed:
add print statements

aditya0by0 committed:
remove print for dataloader phase

aditya0by0 committed:
Merge branch 'fix/directed-to-undirected-graph' into thesis_augmented_gnn

aditya0by0 committed:
gat model pop config if used

to avoid unnecessary kwargs getting passed to GAT

aditya0by0 committed:
gat config

aditya0by0 committed:
add GAT wrapper to init models

aditya0by0 committed:
heads should be divisible by output or hidden channels

aditya0by0 committed:
why in_length is needed ? if n_atom_properties is available

aditya0by0 committed:
fg level edge fix

If two atoms of a FG points to atom(s) belonging to another FG. In this case, only one edge is counted.

aditya0by0 committed:
assert statements for num of edges and nodes

aditya0by0 committed:
bond prop fix

aditya0by0 committed:
num_undir_edges key fix

aditya0by0 commented:
https://wandb.ai/chebai/chebai/runs/f75dyyz3/logs
I trained GAT model (8-heads)  with `GraphPropertyReader` (that is the regular dataset without any augmentation), It achieves **slightly better performance** than Resgated Model with **much less trainable parameters**.

sfluegel05 commented:
That's a good result. Let's sort out the factors involved here:
- data - you said you are not using the augmented dataset, but in the logs I see that the augmented dataset gets loaded and the GAT model uses this data. So what is happening there?
- overfitting - I already mentioned this in #12. The performance increase might stem from a reduced overfitting (indeed, the gap between train and val scores is lower than for https://wandb.ai/chebai/chebai/runs/dlbf4d5s/workspace). So I guess a fair comparison would be one against a ResGated GCN model with the same number of parameters
- other ResGated GCN runs - I did a couple of runs with a ResGated GCN a few weeks ago and got slightly better results than your ResGated and GAT runs (e.g. https://wandb.ai/chebai/chebai/runs/0oksfx9u/logs). I'm not sure what you did differently, but you might want to have a look at that (also, my GCN has 2.7m parameters, yours has 3.6m, but mine has more overfitting - this is a bit contrary to my expectations).

aditya0by0 commented:
> * data - you said you are not using the augmented dataset, but in the logs I see that the augmented dataset gets loaded and the GAT model uses this data. So what is happening there?

You're absolutely right — this was a mistake on my part. I initially intended to say that I trained the GAT model without the augmented **graphs**. Specifically, I used the `GraphPropertyReader` (which used non-augmented graph). 

> * overfitting - I already mentioned this in [Clarify and Fix Usage of Input Channels in Convolution Layers #12](https://github.com/ChEB-AI/python-chebai-graph/issues/12). The performance increase might stem from a reduced overfitting (indeed, the gap between train and val scores is lower than for https://wandb.ai/chebai/chebai/runs/dlbf4d5s/workspace). So I guess a fair comparison would be one against a ResGated GCN model with the same number of parameters

That's a good point. Matching the number of parameters between GAT and ResGated GCN architectures might be challenging due to architectural differences (e.g., GAT uses attention heads, whereas ResGated GCN has gated residual connections). I wonder how can we achieve to make the number parameters of 2 different architecture **exactly** same for comparison.

aditya0by0 commented:
> * other ResGated GCN runs - I did a couple of runs with a ResGated GCN a few weeks ago and got slightly better results than your ResGated and GAT runs (e.g. https://wandb.ai/chebai/chebai/runs/0oksfx9u/logs). I'm not sure what you did differently, but you might want to have a look at that (also, my GCN has 2.7m parameters, yours has 3.6m, but mine has more overfitting - this is a bit contrary to my expectations).

The runs which you did were on **directed** graphs. My runs are on **undirected** graphs. This might be one difference. 

I had did a run on directed graph (i.e dev branch) same as yours few weeks back https://wandb.ai/chebai/chebai/runs/rec4vmfc/overview. 


<p>Below is a comparison of our runs:</p>
<h3>🔬 <strong>Run Comparison (Directed Graph, ResGated GCN)</strong></h3>

Metric | My Run (rec4vmfc) | Your Run (0oksfx9u)
-- | -- | --
Epoch | 199 | 199
Train Loss (epoch) | 0.000864 | 0.000885
Train Macro-F1 | 0.9605 | 0.9603
Train Micro-F1 | 0.9904 | 0.9903
Val Loss (epoch) | 0.02035 | 0.01965
Val Macro-F1 | 0.6821 | 0.6991
Val Micro-F1 | 0.9101 | 0.9111
Global Step | **62,799** | **62,199**


<p>Even though the setups are nearly identical, there are <strong>noticeable differences in metrics and <code inline="">global_step</code></strong>, which might suggest a difference in data size</p>
<p>The discrepancy in <code inline="">global_step</code> (62,799 vs. 62,199) especially indicates that the number of batches per epoch might differ slightly, likely due to subtle changes in how the dataset is loaded or split in different versions of the <code inline="">chebai</code> package.</p>




### 🛠️ **Environment Differences**

| Component      | My Environment | Your Environment |
| -------------- | -------------- | ---------------- |
| **chebai**     | `1.0.0`        | `0.0.2.dev0`     |
| **rdkit**      | `2024.3.6`     | `2024.9.4`       |
| **Graph Type** | Directed       | Directed         |

sfluegel05 commented:
> I wonder how can we achieve to make the number parameters of 2 different architecture **exactly** same for comparison.

I guess exact is not necessary here, within ~200,000 of each other should be enough for a comparison. 

> The discrepancy in global_step (62,799 vs. 62,199) especially indicates that the number of batches per epoch might differ slightly, likely due to subtle changes in how the dataset is loaded or split in different versions of the chebai package.

That's a good point. Since we are not using the exact same dataset and splits, 1 or 2 percent deviation are not surprising. In that case, you should not be too worried about the comparison to my run, but continue with your setup

aditya0by0 committed:
graphpropertyreader should inherit from Data reader

instead of ChemDataReader,
- As no methods from ChemDataReader are resused
- Also it allows avoid unnecessary creation of reader_name/tokens.txt file
- As this token files are creater by encoder in graph repo

aditya0by0 committed:
merge

aditya0by0 committed:
Merge branch 'fix/after_setup_hook' into thesis_augmented_gnn

aditya0by0 committed:
for ring, assigned connected atom ring_size as fg

aditya0by0 committed:
make fg graph code more efficient

aditya0by0 committed:
Merge branch 'dev' into thesis_augmented_gnn

aditya0by0 committed:
add plot option to plot only rdkit molecule using rdkit

aditya0by0 committed:
fix for atoms with no functional group

aditya0by0 committed:
add fg summary to reader

aditya0by0 committed:
assertion for atom belong to more than one fg

aditya0by0 committed:
Merge branch 'fix/optimize_index_encoder' into thesis_augmented_gnn

aditya0by0 committed:
no default values, need to explicitly provided

aditya0by0 committed:
rectify assertion for multiple fg check

aditya0by0 committed:
common token for atoms with no FG

- instead of atom symbol or fg smiles to limit the number of FG tokens

aditya0by0 commented:
https://wandb.ai/chebai/chebai/runs/0snlkito/overview

@sfluegel05 
Please find the run for **Augmented Graphs** with GAT model. 

Only 4 properties are used:
0:"chebai_graph.preprocessing.properties.AtomFunctionalGroup"
1:"chebai_graph.preprocessing.properties.AtomNodeLevel"
2:"chebai_graph.preprocessing.properties.AtomRingSize"
3:"chebai_graph.preprocessing.properties.BondLevel" 


I will compare this with earlier run https://wandb.ai/chebai/chebai/runs/f75dyyz3/logs with GAT on non-augmented graph with usual properties 

<html>
<body>
<!--StartFragment--><html><head></head><body><p>Here is a comparison table between the <strong>GAT without augmented graph</strong> and the <strong>GAT with augmented graph</strong>, based on training/validation metrics and property sets:</p>
<hr>
<h3>📊 <strong>GAT Model Performance Comparison</strong></h3>

Metric | Without Augmented Graph | With Augmented Graph
-- | -- | --
Epoch | 199 | 199
Train Loss (Epoch) | 0.00347 | 0.00455
Train Loss (Step) | 0.00372 | 0.00509
Train Macro-F1 | 0.854 | 0.819
Train Micro-F1 | 0.956 | 0.940
Val Loss (Epoch) | 0.00967 | 0.01333
Val Loss (Step) | 0.00186 | 0.01342
Val Macro-F1 | 0.677 | 0.598
Val Micro-F1 | 0.907 | 0.880
Global Steps | 62,799 | 62,799


### 🔬 **Properties Used**

| Type                    | **Without Augmented Graph (Undirected)**                                                       | **With Augmented Graph (Undirected)**                         |
| ----------------------- | --------------------------------------------------------------------------------- | ------------------------------------------------ |
| **Atom Properties**     | AtomType, NumAtomBonds, AtomCharge, AtomAromaticity, AtomHybridization, AtomNumHs | AtomFunctionalGroup, AtomNodeLevel, AtomRingSize |
| **Bond Properties**     | BondType, BondInRing, BondAromaticity                                             | BondLevel                                        |
| **Additional Features** | RDKit2DNormalized                                                                 | –                                                |

---

aditya0by0 commented:
Resgated _Undirected_ **Augmented Graph** 200 epochs https://wandb.ai/chebai/chebai/runs/at7a5rhr/ 
Resgated _Undirected_ **Augmented Graph** 600 epochs https://wandb.ai/chebai/chebai/runs/slttgfvm/overview

Reference for comparison for Resgated **non-augmented** _Undirected_ graph 200 epochs: https://wandb.ai/chebai/chebai/runs/dlt1iug5/overview 


GAT _Undirected_ **Augmented Graph** 600 epochs https://wandb.ai/chebai/chebai/runs/eehn0d3s/overview

aditya0by0 commented on chebai_graph/preprocessing/reader/augmented_reader.py:
```diff
@@ -0,0 +1,575 @@
+                for atom in connected_atoms:
+                    atom.SetProp("FG", f"RING_{ring_size}")
+
+            else:  # No connected atoms have a ring size which indicates it is simple FG
+                fg_set = {atom.GetProp("FG") for atom in connected_atoms}
+                if not fg_set:
+                    raise ValueError(
+                        "No functional group assigned to atoms in the functional group."
+                    )
+
+                if "" in fg_set and len(fg_set) == 1:
+                    if len(connected_atoms) == 1:
+                        # If there is only one atom and one edge connecting this atom to its fg_atom,
+                        # the functional group will be the symbol of this atom
+                        # This special case is to handle wildcard SMILES Eg. CHEBI:33429
+                        atom = connected_atoms[0]
+                        atom.SetProp("FG", atom.GetSymbol())
+                        self._cnt_fg_using_atom_symbol += 1
+                        flag_mol_has_fg_using_atom_symbol = True
+                    else:
+                        # If there are multiple atoms connected to the functional group, and no atoms have a functional group property/name
+                        # assigned, we assign the functional group as the relevant part of SMILES which belong to the functional group
+                        # Eg. CHEBI:55388, atom idx 2 and 3  have no functional group name, so "[C-]#[C-]" is used
+                        for atom in connected_atoms:
+                            atom.SetProp("FG", fg_smiles)
```
- For single atom, which do not belong to any Functional Group. It will have a FG node connected to it, with the atom's  symbol as its Functional Group property name.

-  If there are multiple atoms connected to the functional group node, and no atoms have a functional group property/name assigned, we assign the functional group as the relevant part of SMILES which belong to the functional group

aditya0by0 replied:
I thought that this might lead to increase in unique Functional Group in our token file, but this not the case.
There are **2533** unique Functional groups in token file out of which **2056** are related to RINGS.

Please find the attachment of the index-token file for function group.
[atom_functional_groups.txt](https://github.com/user-attachments/files/20755489/atom_functional_groups.txt)

aditya0by0 committed:
fix empty node error

aditya0by0 committed:
print statemetns instead rank in pre-processing stage

aditya0by0 committed:
Merge branch 'dev' into thesis_augmented_gnn

aditya0by0 committed:
Revert "common token for atoms with no FG"

This reverts commit 7d6608d8999942923265b01990e52c24883285af.

aditya0by0 committed:
separate out logic of setting fg prop into diff funcs

aditya0by0 committed:
separate frag for each ring in fused ring

aditya0by0 committed:
max_ring_size of atom belonging to  more than one ring in fused ring

aditya0by0 committed:
edges between fg of rings belonging to fused rings

aditya0by0 committed:
3d ◀️  mean position of connected atoms for aug nodes

aditya0by0 committed:
fix huge fused ring issue

aditya0by0 committed:
import data class in init

aditya0by0 committed:
fix fg None error

aditya0by0 committed:
add usual prop for augmented graph with default value

aditya0by0 committed:
set tensor of zeros if None value is given to encoder

aditya0by0 committed:
return None for augmented nodes

aditya0by0 committed:
move base prop class into another base file

aditya0by0 committed:
add wrapper to use mol prop for augmented graph

aditya0by0 committed:
avoid * imports for linters

aditya0by0 committed:
restore aug prop dervied from aug graph to original cls name

aditya0by0 committed:
fix start imports in reader + minor other fix

aditya0by0 committed:
remove lambd exp as suggested by ruff linter

https://docs.astral.sh/ruff/rules/lambda-assignment/

aditya0by0 committed:
remove chebi version

aditya0by0 commented:
### Comparison of Regated Undirected Model (With vs Without Augmented Graph)

_Note_: **Same properties are used in both case**
    **n_conv_layers**: **2** in both case (which means 3 convolutions layers are used)


| Metric                       | [Without Augmented Graph](https://wandb.ai/chebai/chebai/runs/bagd8xd8/overview)     | [With Augmented Graph](https://wandb.ai/chebai/chebai/runs/xy87z6pn/overview)        |
|-----------------------------|-----------------------------|-----------------------------|
| **Epoch**                   | 199                         | 199                         |
| **Train Loss (Epoch)**      | 0.00157                     | 0.00149                     |
| **Train Loss (Step)**       | 0.00196                     | 0.00149                     |
| **Train Macro-F1**          | 0.9251                      | **0.9312**                      |
| **Train Micro-F1**          | 0.9813                      | **0.9821**                      |
| **Val Loss (Epoch)**        | 0.01778                     | 0.01753                     |
| **Val Loss (Step)**         | 0.00975                     | 0.00812                     |
| **Val Macro-F1**            | 0.6628                      | **0.6740**                      |
| **Val Micro-F1**            | 0.9040                      | **0.9062**                      |
| **trainer/global_step**            | 62,799                      | 62,799                      |
| **ResGatedGraphConvNetBase**| 1.2 M                       | 1.2 M                       |
| **Linear Layers**           | 365 K                       | 365 K                       |
| **Final Layer**             | 783 K                       | 783 K                       |

aditya0by0 commented:
### Comparison of GAT Undirected Model (With vs Without Augmented Graph)

_Note_: **Same properties are used in both case**

| Metric                       | [Without Augmented Graph](https://wandb.ai/chebai/chebai/runs/50z1fn5a/overview)     | [With Augmented Graph](https://wandb.ai/chebai/chebai/runs/lipne1mx/overview)        |
|-----------------------------|-----------------------------|-----------------------------|
| **Epoch**                   | 199                         | 199                         |
| **Global Step**             | 62,799                      | 62,799                      |
| **Train Loss (Epoch)**      | 0.00346                     | 0.00353                     |
| **Train Loss (Step)**       | 0.00380                     | 0.00398                     |
| **Train Macro-F1**          | 0.8532                      | **0.8550**                      |
| **Train Micro-F1**          | **0.9559**                      | 0.9550                      |
| **Val Loss (Epoch)**        | 0.00980                     | 0.01016                     |
| **Val Loss (Step)**         | 0.00430                     | 0.00796                     |
| **Val Macro-F1**            | 0.6806                      | **0.6833**                      |
| **Val Micro-F1**            | **0.9061**                      | 0.9060                      |
| **GAT Parameters**          | 622 K                       | 622 K                       |
| **Linear Layers**           | 627 K                       | 627 K                       |
| **Final Layer**             | 783 K                       | 783 K                       |

aditya0by0 commented:
### 🔍 ResGated Graph Comparison (with vs. without Augmented Graph)

**n_conv_layers: 3** in both case (which means 4 convolutions layers are used)
| Metric               | [With Augmented Graph](https://wandb.ai/chebai/chebai/runs/3b7uo82e/overview)      | [Without Augmented Graph](https://wandb.ai/chebai/chebai/runs/7qcd1oyy/overview)   |
|----------------------|---------------------------|----------------------------|
| **Epoch**            | 199                       | 199                        |
| **Train Loss (epoch)** | 0.00150                  | 0.00140                    |
| **Train Loss (step)**  | 0.00203                  | 0.00226                    |
| **Train Macro-F1**     | **0.9275**                   | 0.9271                     |
| **Train Micro-F1**     | 0.9817                   | **0.9830**                     |
| **Validation Loss (epoch)** | 0.01735            | 0.01791                    |
| **Validation Loss (step)**  | 0.00606            | 0.00849                    |
| **Val Macro-F1**        | **0.6732**                   | 0.6695                     |
| **Val Micro-F1**        | 0.9069                   | **0.9070**                     |
| **Global Steps**        | 62,799                   | 62,799                     |
| **ResGatedGraphConvNetBase**| 1.5 M                       | 1.5 M                       |
| **Linear Layers**           | 365 K                       | 365 K                       |
| **Final Layer**             | 783 K                       | 783 K                       |

aditya0by0 commented:
@sfluegel05 

Right now, I can think of two reasons why the **augmented graph** (with the same properties as the non-augmented graph) is not giving a significant improvement in classification performance:

1. **Zeroed-out features for augmented nodes**
   We currently assign a vector of zeros for all properties related to the augmented node. This might be hurting the model, as these nodes contribute structurally but don’t carry informative features during message passing.

2. **Dilution in graph-level pooling**
   The final graph representation is obtained by averaging all node embeddings (atoms + augmented nodes). Since the number of augmented nodes is significantly smaller than the atoms, their signal likely gets diluted. This might explain why the improvement is minimal.

---

### Possible Solution

To address this, we could try **separating the pooled representations**:

* One vector: average of **atom embeddings**
* One vector: average of **augmented node embeddings**

Then, the first linear layer would take as input:
`[atom_repr || augmented_repr || molecular_properties]`
(i.e., shape: `hidden_dim + hidden_dim + n_molecular_properties`)

aditya0by0 committed:
seperate pooling for atom and augmented nodes

One vector: average of atom embeddings
One vector: average of augmented node embeddings
https://github.com/ChEB-AI/python-chebai-graph/pull/2#issuecomment-3006390618

aditya0by0 committed:
hydrogen bond donor/acceptor prop for fg

aditya0by0 committed:
is_alkyl fg prop

aditya0by0 committed:
add unk tokens to prop files

aditya0by0 committed:
base classes for seperate pooling of each type of node

aditya0by0 committed:
Revert "add unk tokens to prop files"

This reverts commit 8353cdd7e8665671f77f9624bc10d3acd5330c78.

aditya0by0 committed:
activation for intermediate lin layers

aditya0by0 committed:
base class for fg node only pooling

aditya0by0 committed:
model wrapper for resgated

aditya0by0 committed:
classes for augmented graph with poolings

aditya0by0 committed:
resgated model separate file

aditya0by0 committed:
add normal regated  without pool

aditya0by0 committed:
add ruff to pre-commit

aditya0by0 committed:
fix ruff errors

aditya0by0 committed:
add ruff to action workflow

aditya0by0 committed:
add gat model

aditya0by0 committed:
Create resgated.yml

aditya0by0 committed:
more class for ablation studies

aditya0by0 committed:
add graph node mask

aditya0by0 committed:
few pooling single only class for ablation study

aditya0by0 committed:
resgated new ablation study classes

aditya0by0 committed:
fg node only  fg nodes with edges only readers

aditya0by0 committed:
augmented readers for ablation studies

aditya0by0 committed:
adapt visualization code for other aug readers

aditya0by0 committed:
Merge branch 'dev' into thesis_augmented_gnn

aditya0by0 committed:
add data classes for ablation readers

aditya0by0 committed:
adapt aug props for ablation data classes

aditya0by0 committed:
doc for chebi dataset class

aditya0by0 committed:
doc for props

aditya0by0 committed:
doc for aug props

aditya0by0 committed:
doc for reader

aditya0by0 committed:
doc for aug reader

aditya0by0 committed:
doc for vis

aditya0by0 committed:
doc for graph collator

aditya0by0 committed:
doc for encoder

aditya0by0 committed:
doc of structures

aditya0by0 committed:
lint and pre-commit version should match

aditya0by0 committed:
doc model base

aditya0by0 committed:
doc for models augmented

aditya0by0 committed:
doc model resgated and gat

aditya0by0 committed:
fix GAT model edge-index and act error

aditya0by0 committed:
data configs for augmented graphs

aditya0by0 committed:
remove mol props comment

aditya0by0 committed:
update model configs

aditya0by0 committed:
NO_FG instead to atom symbol to reduce no of tokens

aditya0by0 committed:
fix RING key error

aditya0by0 committed:
fg detection algo - instead of atom symbol use NO_FG

aditya0by0 committed:
fix: 'int' object has no attribute 'split'

aditya0by0 committed:
add alkyl prop to graph node

aditya0by0 committed:
Revert "fg detection algo - instead of atom symbol use NO_FG"

This reverts commit 331f9417e27803b2edea545b86c48c8a06573948.

aditya0by0 committed:
fix: changed number of augmented nodes (rollback to original)

aditya0by0 committed:
refactor set fg properties logic

aditya0by0 committed:
atomring size not required as its included as functional group

aditya0by0 committed:
restructure properties module

aditya0by0 committed:
add class for props as per node type

aditya0by0 committed:
rearrange prop so that allnode prop type to be first

aditya0by0 committed:
add atom and bond level prop tokens

aditya0by0 committed:
restructure resgated params - add pyg impl to compare

aditya0by0 committed:
add test for grap prop as per node class

aditya0by0 committed:
modify merge method to have diff props for diff nodes

aditya0by0 committed:
use pyg model impl without changing the architecture

- https://github.com/ChEB-AI/python-chebai-graph/issues/12

aditya0by0 committed:
zero padding for atom properties

aditya0by0 committed:
sort as per node prop to same seq as baseline for fair comparison

aditya0by0 committed:
add string-based sorting for `GraphPropAsPerNodeType` too

https://github.com/ChEB-AI/python-chebai-graph/issues/19

aditya0by0 committed:
static gni

aditya0by0 committed:
add as per node class to init

aditya0by0 committed:
add aug rdkit norm prop

aditya0by0 committed:
fix mol props for as per node data cls

aditya0by0 committed:
add few more ablation study aug readers

aditya0by0 committed:
move vis script to results

aditya0by0 committed:
new chebi data class for new readers

aditya0by0 committed:
as node mask for as per node data cls

aditya0by0 committed:
fix hyperparams for gat and resgated

aditya0by0 committed:
gat aug node pool class

aditya0by0 committed:
add fg tokens file

aditya0by0 committed:
classwise f1 score analysis tutorial

aditya0by0 committed:
gatv2-constrainted

aditya0by0 committed:
Merge branch 'thesis_augmented_gnn' of https://github.com/ChEB-AI/python-chebai-graph into thesis_augmented_gnn

aditya0by0 committed:
GATv2 amg pool

aditya0by0 committed:
Merge branch 'dev' into thesis_augmented_gnn

Commit:
add chebi100 config

Commit:
fix chebi100 config

aditya0by0 committed:
static gni no mol props

aditya0by0 committed:
fix gat v2 share weights issue

aditya0by0 committed:
Merge branch 'thesis_augmented_gnn' into feature/gni

aditya0by0 committed:
revert mol props

aditya0by0 committed:
dynamic gni

aditya0by0 committed:
fix reader none error

aditya0by0 committed:
dynamic gni fix

sfluegel05 committed:
allow not setting n_molecule_properties

sfluegel05 committed:
Merge branch 'thesis_augmented_gnn' of https://github.com/ChEB-AI/python-chebai-graph into thesis_augmented_gnn

aditya0by0 committed:
add configs for aug

aditya0by0 committed:
set device for random initiated tensors

aditya0by0 committed:
add padding with zeros or randomness for node and edges

aditya0by0 committed:
for dynamic randomness set option for padding existing features with random features

aditya0by0 committed:
if padding applied, create separate data.pt file

sfluegel05 committed:
Merge branch 'thesis_augmented_gnn' of https://github.com/ChEB-AI/python-chebai-graph into thesis_augmented_gnn

sfluegel05 committed:
add ChEBI100 aug class

aditya0by0 committed:
minor changes for gni

aditya0by0 committed:
Merge branch 'thesis_augmented_gnn' of https://github.com/ChEB-AI/python-chebai-graph into thesis_augmented_gnn

aditya0by0 committed:
lint base file f

aditya0by0 committed:
include zero padding random init reader static method

sfluegel05 committed:
add optional batch norm

sfluegel05 committed:
fix merge conflict
